applications/machine_learning/src/modules/usb_state.c

```diff
@@ -8,15 +8,108 @@
 
 #define MODULE usb_state
 #include <caf/events/module_state_event.h>
+#include <caf/events/power_manager_event.h>
 
 #include <logging/log.h>
 LOG_MODULE_REGISTER(MODULE, CONFIG_ML_APP_USB_STATE_LOG_LEVEL);
 
+/** @brief Peer states. */
+enum usb_state {
+	USB_STATE_DISCONNECTED, /**< Cable is not attached. */
+	USB_STATE_POWERED,      /**< Cable attached but no communication. */
+	USB_STATE_ACTIVE,       /**< Cable attached and data is exchanged. */
+	USB_STATE_SUSPENDED,    /**< Cable attached but port is suspended. */
+	USB_STATE_ERROR         /**< USB error */
+};
+
+
+/* Current state of the USB */
+static enum usb_state current_state = USB_STATE_DISCONNECTED;
+
+
+static void device_status(enum usb_dc_status_code status, const uint8_t *param)
+{
+	static enum usb_state before_suspend;
+	enum usb_state new_state = current_state;
+
+	switch (status) {
+	case USB_DC_ERROR:
+		new_state = USB_STATE_ERROR;
+		module_set_state(MODULE_STATE_ERROR);
+		break;
+	case USB_DC_RESET:
+		__ASSERT_NO_MSG(current_state != USB_STATE_DISCONNECTED);
+		new_state = USB_STATE_ACTIVE;
+		break;
+	case USB_DC_CONNECTED:
+		new_state = USB_STATE_POWERED;
+		break;
+	case USB_DC_CONFIGURED:
+		__ASSERT_NO_MSG(current_state != USB_STATE_DISCONNECTED);
+		new_state = USB_STATE_ACTIVE;
+		break;
+	case USB_DC_DISCONNECTED:
+		new_state = USB_STATE_DISCONNECTED;
+		break;
+	case USB_DC_SUSPEND:
+		if (current_state == USB_STATE_DISCONNECTED) {
+			/* Due to the way USB driver and stack are written
+			 * some events may be issued before application
+			 * connect its callback.
+			 * We assume that device was powered.
+			 */
+			current_state = USB_STATE_POWERED;
+			LOG_WRN("USB suspended while disconnected");
+		}
+		before_suspend = current_state;
+		new_state = USB_STATE_SUSPENDED;
+		LOG_WRN("USB suspend");
+		break;
+	case USB_DC_RESUME:
+		__ASSERT_NO_MSG(current_state != USB_STATE_DISCONNECTED);
+		if (current_state == USB_STATE_SUSPENDED) {
+			new_state = before_suspend;
+			LOG_WRN("USB resume");
+		}
+		break;
+	case USB_DC_INTERFACE:
+	case USB_DC_SET_HALT:
+	case USB_DC_CLEAR_HALT:
+	case USB_DC_SOF:
+		/* Ignore */
+		break;
+	case USB_DC_UNKNOWN:
+	default:
+		/* Should not happen */
+		__ASSERT_NO_MSG(false);
+		break;
+	}
+
+	if (current_state != new_state) {
+		current_state = new_state;
+		switch (new_state) {
+		case USB_STATE_ERROR:
+		case USB_STATE_DISCONNECTED:
+			power_manager_restrict(MODULE_IDX(MODULE), POWER_MANAGER_LEVEL_MAX);
+			break;
+		case USB_STATE_POWERED:
+		case USB_STATE_ACTIVE:
+			power_manager_restrict(MODULE_IDX(MODULE), POWER_MANAGER_LEVEL_ALIVE);
+			break;
+		case USB_STATE_SUSPENDED:
+			power_manager_restrict(MODULE_IDX(MODULE), POWER_MANAGER_LEVEL_SUSPENDED);
```

Comment:
Should we also force power down in similar way as we do it in usb_state_pm ?

Reply:
We still have NCSDK-10437 to handle that

Reply:
I meant more like scheduling a work item and calling force_power_down from it (in the same way as we do it for nRF Desktop's usb_state_pm module). I remember that we still need to improve it, but maybe we should keep it consistent for now.

Reply:
Ok. I will include such a functionality here.

Reply:
Done